clock/mock.go

```diff
@@ -0,0 +1,420 @@
+		}
+	}
+	if len(traps) == 0 {
+		return
+	}
+	c.complete.Add(len(traps))
+	m.mu.Unlock()
+	for _, t := range traps {
+		go t.catch(c)
+	}
+	c.complete.Wait()
+	m.mu.Lock()
+}
+
+// Advance moves the clock forward by d, triggering any timers or tickers.  Advance will wait for
+// tick functions of tickers created using NewContextTicker to complete before returning from
+// Advance. If multiple timers or tickers trigger simultaneously, they are all run on separate go
+// routines.
+func (m *Mock) Advance(d time.Duration) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.advanceLocked(d)
+}
+
+func (m *Mock) advanceLocked(d time.Duration) {
+	if m.advancing {
+		panic("multiple simultaneous calls to Advance not supported")
+	}
+	m.advancing = true
+	defer func() {
+		m.advancing = false
+	}()
+
+	fin := m.cur.Add(d)
+	for {
+		// nextTime.IsZero implies no events scheduled
+		if m.nextTime.IsZero() || m.nextTime.After(fin) {
+			m.cur = fin
+			return
+		}
+
+		if m.nextTime.After(m.cur) {
+			m.cur = m.nextTime
+		}
+
+		wg := sync.WaitGroup{}
+		for i := range m.nextEvents {
+			e := m.nextEvents[i]
+			t := m.cur
+			wg.Add(1)
+			go func() {
+				e.fire(t)
+				wg.Done()
+			}()
+		}
+		// release the lock and let the events resolve.  This allows them to call back into the
+		// Mock to query the time or set new timers.  Each event should remove or reschedule
+		// itself from nextEvents.
+		m.mu.Unlock()
+		wg.Wait()
+		m.mu.Lock()
+	}
+}
+
+// Set the time to t.  If the time is after the current mocked time, then this is equivalent to
+// Advance() with the difference.  You may only Set the time earlier than the current time before
+// starting tickers and timers (e.g. at the start of your test case).
+func (m *Mock) Set(t time.Time) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if t.Before(m.cur) {
+		// past
+		if !m.nextTime.IsZero() {
+			panic("Set mock clock to the past after timers/tickers started")
+		}
+		m.cur = t
+		return
+	}
+	// future, just advance as normal.
+	m.advanceLocked(t.Sub(m.cur))
+}
+
+// Trapper allows the creation of Traps
+type Trapper struct {
+	// mock is the underlying Mock.  This is a thin wrapper around Mock so that
+	// we can have our interface look like mClock.Trap().NewTimer("foo")
+	mock *Mock
+}
+
+func (t Trapper) NewTimer(tags ...string) *Trap {
+	return t.mock.newTrap(clockFunctionNewTimer, tags)
+}
+
+func (t Trapper) TimerStop(tags ...string) *Trap {
+	return t.mock.newTrap(clockFunctionTimerStop, tags)
+}
+
+func (t Trapper) TimerReset(tags ...string) *Trap {
+	return t.mock.newTrap(clockFunctionTimerReset, tags)
+}
+
+func (t Trapper) NewContextTicker(tags ...string) *Trap {
+	return t.mock.newTrap(clockFunctionNewContextTicker, tags)
+}
+
+func (t Trapper) ContextTickerWait(tags ...string) *Trap {
+	return t.mock.newTrap(clockFunctionContextTickerWait, tags)
+}
+
+func (m *Mock) Trap() Trapper {
+	return Trapper{m}
+}
+
+func (m *Mock) newTrap(fn clockFunction, tags []string) *Trap {
```

Comment:
Is it possible to add a trap for a tag that does not exist? If so, can this be caught early and immediately call `t.Fatal()`?